lib/internal/modules/esm/fetch.js

```diff
@@ -0,0 +1,310 @@
+'use strict';
+const {
+  globalThis,
+  ArrayFrom,
+  ArrayPrototypeJoin,
+  ArrayPrototypePush,
+  ArrayPrototypeSplice,
+  Promise,
+  PromisePrototypeThen,
+  PromiseResolve,
+  SafeMap,
+  StringPrototypeEndsWith,
+  StringPrototypePadStart,
+  StringPrototypeSlice,
+  StringPrototypeSplit,
+  StringPrototypeStartsWith,
+} = primordials;
+const {
+  Buffer: {
+    concat: BufferConcat
+  }
+} = globalThis;
+const {
+  ERR_NETWORK_IMPORT_DISALLOWED,
+  ERR_NETWORK_IMPORT_BAD_RESPONSE,
+} = require('internal/errors').codes;
+const { fileURLToPath, URL } = require('internal/url');
+const path = require('path');
+
+/**
+ * @typedef CacheEntry
+ * @property resolvedHREF {Promise<string> | string}
+ * @property headers {Record<string, string>}
+ * @property body {Promise<Buffer> | Buffer}
+ */
+
+/**
+ * Only for GET requests, other requests would need new Map
+ * HTTP cache semantics keep diff caches
+ *
+ * Maps HREF to pending cache entry
+ * @type {Map<string, Promise<CacheEntry> | CacheEntry>}
+ */
+const cacheForGET = new SafeMap();
+
+
+/**
+ * Lazily load these to avoid snapshot issues
+ */
+let HTTPSAgent;
+function HTTPSGet(url, opts) {
+  const https = require('https');
+  HTTPSAgent ??= new https.Agent({
+    keepAlive: true
+  });
+  return https.get(url, {
+    agent: HTTPSAgent,
+    ...opts
+  });
+}
+
+let HTTPAgent;
+function HTTPGet(url, opts) {
+  const http = require('http');
+  HTTPAgent ??= new http.Agent({
+    keepAlive: true
+  });
+  return http.get(url, {
+    agent: HTTPAgent,
+    ...opts
+  });
+}
+
+function dnsLookup(name, opts) {
+  // eslint-disable-next-line no-func-assign
+  dnsLookup = require('dns/promises').lookup;
+  return dnsLookup(name, opts);
+}
+
+function createBrotliDecompress() {
+  // eslint-disable-next-line no-func-assign
+  createBrotliDecompress = require('zlib').createBrotliDecompress;
+  return createBrotliDecompress();
+}
+
+function createUnzip() {
+  // eslint-disable-next-line no-func-assign
+  createUnzip = require('zlib').createUnzip;
+  return createUnzip();
+}
```

Comment:
```suggestion
let zlib;
function createBrotliDecompress() {
  zlib ??= require('zlib');
  // eslint-disable-next-line no-func-assign
  createBrotliDecompress = zlib.createBrotliDecompress;
  return createBrotliDecompress();
}

function createUnzip() {
  zlib ??= require('zlib');
  // eslint-disable-next-line no-func-assign
  createUnzip = zlib.createUnzip;
  return createUnzip();
}
```

`require('zlib')` is used across `createBrotliDecompress` and `createUnzip`. Cache the `require` **might** improve the performance here?

Reply:
This looks good to me 🙂

@bmeck would this be kosher for the V8 issue you mentioned that necessitates not eagerly requiring these?

Reply:
yep, this is fine; the issue is only when `require` is called in the snapshot which this does not do.